### ISSUE TITLE
Read URL from package.json

### ISSUE THREAD

**Issue by egoist** [feross/thanks#2]
Feel free to close it if this tool is only made for the maintainers and projects listed in https://github.com/feross/thanks/blob/master/index.js

The __feature request__ is to support __`"thanks": "url"`__ field in `package.json` so that all packages with this field set can be displayed in the _thanks_ table. 😄

**Comment by feross** [comment]
This is a neat idea.

**Comment by noffle** [comment]
Decentralize! :heart:

**Comment by zkat** [comment]
One thing worth noting is that you can already add arbitrary contact info for individual `author`s, which might also be preferable. See [the expanded contact info I use](https://github.com/zkat/npx/blob/latest/package.json#L33-L37) for my projects, which already integrates with [`weallbehave`](https://github.com/wealljs/weallbehave#npm-repo) for figuring out who to contact. You can probably reuse this field easily enough, or use it for lookup.

**Comment by egoist** [comment]
Yep something like `author.donation` `author.patreon` `author.openCollective` looks good to me too 😄

**Comment by qubyte** [comment]
Using the expanded author info is intriguing, particularly when using the `url` field. If that URL served a document with a link or anchor tag for donations, this'd work well with indie web:

https://indieweb.org/payment
http://microformats.org/wiki/rel-payment

**Comment by marijnh** [comment]
This'd also address the problem of the burden of keeping the donation link data accurate being on you (if this takes off, I can imagine people submitting PRs to associates some popular developer with a patreon they control)

**Comment by olstenlarck** [comment]
I'm both for `thanks` field and inside `author`. :tada: :taco:

**Comment by btkostner** [comment]
The `url` field is designed to be a homepage link, which may or may not be a donation link, but it would be best to keep them seperate. While the `thanks` field would work for this project, I personally think we should use the `donation` field. It's general enough that it can be used in other projects, and hopefully eventually be a part of the `package.json` spec. Secondly, allowing `author.donation` has the added benifit of being able to have multiple donation links for a package.

Down the line if we need more information, the `donation` field could be expanded to an object, similar to the `repository` field.
```json
{
  "donation": {
    "type": "patreon",
    "url": "..."
  }
}
```

**Comment by pubkey** [comment]
if we put the donation-data inside of a `.thanks.json`, this logic can later be reused with non-npm-projects.

**Comment by qubyte** [comment]
@btkostner To be clear, my suggestion does not change that `url` is for a homepage. The rel-payment microformat (see links) is a way of discovering a donation URL _within_ a page (either as a link tag or an anchor tag). This means that anyone/anything visiting a homepage with one will easily be able to discover a way to donate.

**Comment by qubyte** [comment]
(something explicit in the package file would also be good, but this would be nice to have as another mode of donation discovery)

**Comment by adrianheine** [comment]
I would love to see @qubyte's suggestion implemented, both for ideological reasons and to ease maintenance. This should only work over HTTPS and validate certs, though. Explicit info from `package.json` should have precedence, though.

**Comment by qubyte** [comment]
@adrianheine Agreed on all counts. I'm putting a module together now which does this. I may open move the suggestion as a separate issue so I don't hijack this comment thread any further.

**Comment by xdamman** [comment]
I like @btkostner but I think the type shouldn't be used to say "opencollective" or "patreon" because those can be easily figured out from the url.

Instead the type could be: `url` or `bitcoin` or `ethereum`.
e.g.

```
{
  "donation": {
    "type": "url",
    "url": "https://opencollective.com/webpack"
  }
}
```

```
{
  "donation": {
    "type": "bitcoin",
    "address": "14MEpy5a9MwDZa9CUzrfDhTU8dy2KKJ5mU"
  }
}
```

Issue: we cannot really use `url` anymore, hence `address` or maybe `value`?

**Comment by xdamman** [comment]
Alternatively, we could use the key as the type:

```
{
  "donation": {
    "url": "https://opencollective.com/vuejs",
    "opencollective": "vuejs",
    "btc": "14MEpy5a9MwDZa9CUzrfDhTU8dy2KKJ5mU",
    "bch": "15gftB3fwumFRWGWFhVzTgc4nhM5w1F2Tu",
    "eth": "0x3411356C1f0Bf5D859464eD2AC54DD2C222519B7",
    "ltc": "LUcHis3B8SFtEeZtuCaZoqsyN9XFAKmbCP"
  }
}
```

(those are real values for [VueJS](https://vuejs.org/support-vuejs/))

**Comment by adrianheine** [comment]
`donation` could be an array with each entry having `type` (`eth`, `btc`, `opencollective`, `url`) and `value` (?) fields.

**Comment by xdamman** [comment]
That looks much more verbose. An array of just types and values really sounds like an object with attribute and value no? Also, an object makes it much easier to quickly test if a project supports a certain donation method (`if (pkg.donation.btc) {}`)

The only reason to have an array would be if we want to store more data (e.g. current balance?). But I don't think that this is the place for it.

**Comment by marijnh** [comment]
There's an argument to be made for allowing both per-author and per-project links. A multi-author project might want to provide separate links or a single one (say, to an open collective), depending on how it's organized.

**Comment by adrianheine** [comment]
@xdamman You can have multiple values for a specific protocol, you can have multi-dimensional values for a protocol, and you emphasize that the protocol is not from a small, fixed set.

**Comment by adrianheine** [comment]
@marijnh I imagine the flow like this:
1. Check if there is a hard-coded value for the project
2. Otherwise, check if the project has a `donation` value in `package.json`
3. Otherwise, check if there is some standalone file like maybe `thanks.json` or `donations.json`?
4. Otherwise, check if the project's homepage has a `payment` `rel`
5. Otherwise, check if there is a hard-coded value for the npmjs.com maintainers
6. Otherwise, check if there is a `donation` value in the `author` members in `package.json`
7. Otherwise, check if the author's homepages have a `payment` `rel`

I think 3 and 7 would be the preferred ways, 1, 5 deprecated and 2, 4, 6 totally acceptable.

**Comment by bumi** [comment]
I was experimenting with a similar idea a while ago for ruby gems ([bundler thankyou](https://github.com/bumi/bundler-thankyou#bundle-thankyou---simple-donation-system-for-ruby-gems-and-other-package-managers)). My goal was to use bitcoin as a system to transfer some tip directly to all dependencies (one payment to all) 
I liked the idea of transferring value directly and NOT require a service provider (like paypal, opencollective, etc.) (service providers could still offer services for either party (payee or recipient))

Having the donation data in the package description (package.json, etc.) allows us to reuse the current trusted infrastructure, it is super easy for the package maintainer and donations can be defined for every version (for example: ruby on rails could donate the "thanks" for v 4.x to railsgirls) 

I like the simple key/value proposal by @xdamman 👍 
IMO the goal should be to have a minimal, small standard and maintainer should be always in full control over the entry. No third party system should be assumed. This gives everybody the choice and could allow tools and services evolve for either party and all different package managers.

**Comment by balupton** [comment]
Thought I'd jump in with what we do with the Bevry ecosystem (around 300 packages).

We use [boundation](https://github.com/bevry/boundation) to generate and update projects, this injects [the following package.json code](https://github.com/bevry/boundation/blob/master/package.json#L17-L43)

```
  "badges": {
    "list": [
      "travisci",
      "npmversion",
      "npmdownloads",
      "daviddm",
      "daviddmdev",
      "---",
      "patreon",
      "opencollective",
      "flattr",
      "paypal",
      "bitcoin",
      "wishlist",
      "---",
      "slackin"
    ],
    "config": {
      "patreonUsername": "bevry",
      "opencollectiveUsername": "bevry",
      "flattrUsername": "balupton",
      "paypalURL": "https://bevry.me/paypal",
      "bitcoinURL": "https://bevry.me/bitcoin",
      "wishlistURL": "https://bevry.me/wishlist",
      "slackinURL": "https://slack.bevry.me"
    }
  },
```

Which is rendered in the READMEs using [projectz](https://github.com/bevry/projectz), which forwards that field to [badges](https://github.com/bevry/badges) to get the rendered badges.

There was plan so that Projectz will pull in the sponsors, and render them automatically in the readme. Currently it just shows those in the "sponsors" package.json field (similar to authors, contributors, maintainers). However, I never got around to finishing it: https://github.com/bevry/projectz/issues/80

**Comment by pastudan** [comment]
I had a similar idea, and found this thread as I was halfway through building a similar tool. I built a platform for handling software donations, [givethanks.app](https://givethanks.app), and a specification for adding donation metadata to your `package.json`, the [Donate Spec](https://github.com/ThanksApp/donate-spec), I'd love to get everyone's thoughts, particularly the spec.

While developing the spec, I tried to keep everything as generic as possible. As @btkostner said [above](https://github.com/feross/thanks/issues/2#issuecomment-364489098), I think this has a higher chance of getting adopted into the main package.json spec if we use `donate` instead of `thanks` and keep the spec as platform agnostic as possible

Feross' `thanks` is a great utility, and I would love to make a PR that consumes the spec after getting some community input. That way, authors won't need to continue making PRs into this repo to get listed.

**Comment by pastudan** [comment]
@tunnckoCore yeah, you can actually support multiple platforms _and_ multiple recipients!

Check out the 3rd example here: https://github.com/ThanksApp/donate-spec#3-multiple-platforms

**Comment by tunnckoCore** [comment]
Oooh, got it. I'm the recipient, yea :laughing:

**Comment by balupton** [comment]
I've added support for Buy Me A Coffee, Liberapay, Thanks App, and Boost Lab to [badges](https://github.com/bevry/badges). So you can generate the donation buttons in your READMEs using [projects](https://github.com/bevry/projectz).

A good next step would be for projectz to read the "badges" config and the "donate" config and do any appropriate merging of the two, so the two properties can generate each other.

**Comment by pastudan** [comment]
@balupton yeah, merging those two would be awesome! You seem to have started standardized list of funding platforms. That's something that I hinted at toward the bottom of the spec, but it would be nice to have a definitive list somewhere. i.e, should it be always lowercase `paypal`, stylized `PayPal`, or title case `Paypal`?

**Comment by balupton** [comment]
@pastudan perhaps we can do a trimmed down version of my unfinished [sponsored](https://github.com/bevry/sponsored) project, where people can signup, list all their donation platforms, and it spurts out the `donate` and `badges` config for them to use, as well as a nice profile page that they can link to

It could even be used like so with the `donate` spec:

```
 "donate": {
    "recipients": ["balupton", "pastudan"]
 },
```

Which then calls say `https://sponsored.app/recipient/balupton.json` and `https://sponsored.app/recipient/pastudan.json` to fetch their latest and most accurate donate platform spec config.

Or with weighting it could do:

```
 "donate": {
    "recipients": [{id: "balupton", "weight": 50}, {id: "pastudan", "weight": 50}]
 },
```

Perhaps even the service could import people's github and npm packages, to allow a always up to date way of generating the appropriate spec info, without needing to specify it in the packages themselves. For instance, I maintain about 200 packages, and not everyone will be prepared to adopt the complex automated setup that I do, as detailed a few posts earlier.

Later on, if anyone is interested, then it could even be used to add the dynamic features that [sponsored](https://github.com/bevry/sponsored) was meant to have

**Comment by balupton** [comment]
Adding in @jing-c from https://boost-lab.app to the discussion.

**Comment by martinszy** [comment]
Hey, what about civic.json? http://open.dc.gov/civic.json/
Is it possible to consider also expanding that standard?
